File: research/models/meta_regime.py
"""Meta-regime detection for AH-HMM — mirrors ``src/tools/finance/markov-distribution/meta-regime.ts``."""

from __future__ import annotations

from typing import Literal, Sequence, TypedDict

import numpy as np

Environment = Literal["low_uncertainty", "high_uncertainty"]


class MetaRegimeFitResult(TypedDict):
    environment_sequence: list[Environment]
    current_environment: Environment | None
    threshold_used: float


class MetaRegimeDetector:
    """Classifies market environment as low_uncertainty or high_uncertainty."""

    def __init__(self, vol_window: int = 20, high_vol_threshold: float = 0.75) -> None:
        self.vol_window = vol_window
        self.high_vol_threshold = high_vol_threshold
        self._threshold: float | None = None
        self._fitted = False

    @property
    def fitted(self) -> bool:
        return self._fitted

    @property
    def threshold(self) -> float | None:
        return self._threshold

    def fit(self, returns: Sequence[float]) -> MetaRegimeFitResult:
        arr = np.asarray(returns, dtype=float)
        n = len(arr)

        rolling_vol = np.full(n, np.nan)
        for i in range(self.vol_window - 1, n):
            window = arr[i - self.vol_window + 1 : i + 1]
            rolling_vol[i] = float(np.std(window, ddof=1))

        valid = rolling_vol[~np.isnan(rolling_vol)]
        if len(valid) == 0:
            self._threshold = 0.02
        else:
            self._threshold = float(np.percentile(valid, self.high_vol_threshold * 100))

        env_seq: list[Environment] = []
        for i in range(n):
            if np.isnan(rolling_vol[i]) or rolling_vol[i] <= self._threshold:
                env_seq.append("low_uncertainty")
            else:
                env_seq.append("high_uncertainty")

        self._fitted = True
        return {
            "environment_sequence": env_seq,
            "current_environment": env_seq[-1] if env_seq else None,
            "threshold_used": self._threshold,
        }

File: src/tools/finance/markov-distribution/meta-regime.ts
/** Mirrors `research/models/meta_regime.py`. */

export type Environment = 'low_uncertainty' | 'high_uncertainty';

export interface MetaRegimeFitResult {
  environment_sequence: Environment[];
  current_environment: Environment | undefined;
  threshold_used: number;
}

export class MetaRegimeDetector {
  readonly vol_window: number;
  readonly high_vol_threshold: number;
  private _threshold: number | undefined;
  private _fitted = false;

  constructor(vol_window = 20, high_vol_threshold = 0.75) {
    this.vol_window = vol_window;
    this.high_vol_threshold = high_vol_threshold;
  }

  get fitted(): boolean {
    return this._fitted;
  }

  get threshold(): number | undefined {
    return this._threshold;
  }

  fit(returns: number[]): MetaRegimeFitResult {
    const rollingVol = Array(returns.length).fill(Number.NaN) as number[];

    for (let i = this.vol_window - 1; i < returns.length; i++) {
      const window = returns.slice(i - this.vol_window + 1, i + 1);
      rollingVol[i] = sampleStd(window);
    }

    const valid = rollingVol.filter(v => !Number.isNaN(v));
    this._threshold = valid.length === 0 ? 0.02 : percentile(valid, this.high_vol_threshold * 100);

    const environment_sequence = rollingVol.map(vol => (
      Number.isNaN(vol) || vol <= this._threshold! ? 'low_uncertainty' : 'high_uncertainty'
    ));

    this._fitted = true;
    return {
      environment_sequence,
      current_environment: environment_sequence.at(-1),
      threshold_used: this._threshold,
    };
  }
}

function sampleStd(values: number[]): number {
  if (values.length <= 1) return Number.NaN;
  const mean = values.reduce((sum, value) => sum + value, 0) / values.length;
  const variance = values.reduce((sum, value) => sum + (value - mean) ** 2, 0) / (values.length - 1);
  return Math.sqrt(variance);
}

function percentile(values: number[], percentileRank: number): number {
  const sorted = [...values].sort((a, b) => a - b);
  if (sorted.length === 1) return sorted[0];

  const index = (percentileRank / 100) * (sorted.length - 1);
  const lower = Math.floor(index);
  const upper = Math.ceil(index);
  const weight = index - lower;

  return sorted[lower] + (sorted[upper] - sorted[lower]) * weight;
}
